### PR TITLE
Add ExternalCallLib which checks revert data for malicious data

### LINKED ISSUE #1973: Wrap external view calls in LinearPool
LinearPools call `view` functions in external contracts in order to get the wrapped token rate. This is dangerous because of the `LinearPoolRebalancer`, which uses queries to manage the Pool. A malicious or vulnerable rate provider could revert with a revert reason that matches the one used in `BasePool`'s query mechanism, causing the rebalancer to perform an incorrect operation, potentially resulting in theft of Pool funds. Note that no Pool currently deployed is affected by this since Aave's system is not vulnerable- this is so far simply a hypothetical issue.

In order to prevent this, we need to wrap all external calls in a `try/catch` block and prevent a malicious revert (one that will be caught by `BasePool`'s queries) from bubbling up: an easy way to achieve this would be to bubble up a different reason. We could also perform the call manually insted of using `try/catch`.

Once we have #1941, we can have custom detectors that check that no codepaths result in external calls, except those that are covered by the guard described above.

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

This PR introduces the `ExternalCallLib` which exposes a function to check if revert data from external calls will be picked up by BasePool's `_queryAction` function.

~~Note that this currently doesn't fix the issue described in #1973. We must also protect the error signature used by `queryBatchSwap` as well do to this. Currently we only protect joins and exits but not swaps.~~ Fixed now.

We should also include a test of this protection when used within a pool.

## Type of change

- [ ] Bug fix 
- [x] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [x] The diff is legible and has no extraneous changes
- [x] Complex code has been commented, including external interfaces
- [x] Tests are included for all code paths
- [x] The base branch is either `master`, or there's a description of how to merge

## Issue Resolution

closes #1973

### CODE REVIEW

File: pkg/pool-utils/contracts/test/MaliciousQueryReverter.sol [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    function setRevertMaliciously(bool enabled) external {
+        revertMaliciously = enabled;
+    }
+
+    function spoofJoinExitQueryRevert() public pure {
+        uint256[] memory tokenAmounts = new uint256[](2);
+        tokenAmounts[0] = 1;
+        tokenAmounts[1] = 2;
+
+        uint256 bptAmount = 420;
```
Comment (nventuro) [inline pkg/pool-utils/contracts/test/MaliciousQueryReverter.sol:29]: _the number_

File: pkg/pool-linear/test/AaveLinearPool.test.ts [hunk 2]
```diff
@@ -82,15 +85,27 @@ describe('AaveLinearPool', function () {
...
+        // We now double the reserve's normalised income to change the exchange rate to 2:1
+        await mockLendingPool.setReserveNormalizedIncome(bn(2e27));
+        expect(await pool.getWrappedTokenRate()).to.be.eq(fp(2));
+      });
+    });
+
+    context('when Aave reverts maliciously', () => {
+      sharedBeforeEach('make Aave lending pool start reverting', async () => {
+        await mockLendingPool.setRevertMaliciously(true);
+      });
```
Comment (nventuro) [inline pkg/pool-linear/test/AaveLinearPool.test.ts:104]: It'd be good to also have a rebalancer test that fails if the lending pool is malicious. 

Also, it'd be great if we tested both swap and joinexit queries. I know we test both at the lib, but it shouldn't be too much work to have both, and it'll cover all cases.

File: pkg/pool-utils/test/ExternalCallLib.test.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    maliciousReverter = await deploy('MaliciousQueryReverter');
+    caller = await deploy('MockExternalCaller', { args: [maliciousReverter.address] });
+  });
+
+  describe('when calling into a malicious contract in a swap', () => {
+    context('when call is protected', () => {
+      it('reverts with MALICIOUS_QUERY_REVERT', async () => {
+        await expect(caller.protectedSwapExternalCall()).to.be.revertedWith('MALICIOUS_QUERY_REVERT');
+      });
+    });
```
Comment (nventuro) [inline pkg/pool-utils/test/ExternalCallLib.test.ts:21]: We're missing tests where we check we bubble up the original revert reason for non malicious callees. The bubble up tests we have here are all of unprotected calls, meaning our bubble up code is not being executed.

**Review by nventuro — APPROVED** [review]
Very clean, loved the approach!

File: pkg/pool-utils/contracts/test/MaliciousQueryReverter.sol (hunk 1 above)
Comment (TomAFrench) [reply pkg/pool-utils/contracts/test/MaliciousQueryReverter.sol:43]: It's like leaving known bugs in a commit for audit, gotta know if the review is thorough. :smile:

**Pull request merged**